### PR TITLE
[App] Add CloudMultiProcessBackend to run an children App within the Flow in the cloud

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR enables to run a VSCode component in the cloud and a local App within it: https://www.loom.com/share/85cd796fdcd14d08b0a2f6474b8574b6

Fixes #\<issue_number>

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: src/lightning_app/utilities/port.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            break
+
+    client.lightningapp_instance_service_update_lightningapp_instance(
+        project_id=project_id,
+        id=lit_app.id,
+        body=AppinstancesIdBody(name=lit_app.name, spec=lit_app.spec),
+    )
+
+    if not found_nc:
+        raise RuntimeError("No available port was found. Please, contact the Lightning Team.")
```
Comment (lantiga) [inline src/lightning_app/utilities/port.py:46]:
```suggestion
        raise RuntimeError("No available port was found. Please open an issue at https://github.com/lightning-AI/lightning/issues")
```

File: src/lightning_app/utilities/port.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
```
Comment (lantiga) [inline src/lightning_app/utilities/port.py:89]:
```suggestion
        raise RuntimeError("No available port was found. Please open an issue at https://github.com/lightning-AI/lightning/issues.")
```

File: src/lightning_app/utilities/port.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+    # Note: This is required for the framework to know we need to use the CloudMultiProcessRuntime.
+    os.environ["APP_SERVER_HOST"] = f"https:/{found_nc.host}"
```
Comment (lantiga) [inline src/lightning_app/utilities/port.py:49]: should it be `httpsL://`?

**Review by lantiga — APPROVED** [review]
Looks good!

File: src/lightning_app/utilities/port.py (hunk 3 above)
Comment (tchaton) [reply src/lightning_app/utilities/port.py]: Done !

File: src/lightning_app/CHANGELOG.md [hunk 4]
```diff
@@ -11,6 +11,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added the CLI command `lightning run model` to launch a `LightningLite` accelerated script ([#15506](https://github.com/Lightning-AI/lightning/pull/15506))
 
 
+- Added the CloudMultiProcessBackend to run an children App within the Flow in the cloud ([#15800](https://github.com/Lightning-AI/lightning/pull/15800))
```
Comment (ethanwharris) [inline src/lightning_app/CHANGELOG.md:14]:
```suggestion
- Added a CloudMultiProcessBackend which enables running a child App from within the Flow in the cloud ([#15800](https://github.com/Lightning-AI/lightning/pull/15800))
```

File: src/lightning_app/utilities/port.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    """Make a request to the cloud controlplane to find a disabled port of the flow, enable it and return it."""
+
+    app_id = os.getenv("LIGHTNING_CLOUD_APP_ID", None)
+    project_id = os.getenv("LIGHTNING_CLOUD_PROJECT_ID", None)
+    enable_multiple_works_in_default_container = bool(int(os.getenv("ENABLE_MULTIPLE_WORKS_IN_DEFAULT_CONTAINER", "0")))
+
+    if not app_id or not project_id or not enable_multiple_works_in_default_container:
+        return default_port
+
+    assert project_id
```
Comment (ethanwharris) [inline src/lightning_app/utilities/port.py:19]: Can be deleted as the above condition will catch it:
```suggestion
```

File: src/lightning_app/utilities/port.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+    client = LightningClient()
+    list_apps_resp = client.lightningapp_instance_service_list_lightningapp_instances(project_id=project_id)
+    lit_app: Optional[Externalv1LightningappInstance] = None
+
+    for lightningapp in list_apps_resp.lightningapps:
+        if lightningapp.id == app_id:
+            lit_app = lightningapp
+
+    assert lit_app
```
Comment (ethanwharris) [inline src/lightning_app/utilities/port.py:29]:
```suggestion
    if not lit_app:
        raise RuntimeError(
            "App was not found. Please open an issue at https://github.com/lightning-AI/lightning/issues."
        )
```

File: src/lightning_app/utilities/port.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+def enable_port() -> V1NetworkConfig:
+    """Make a request to the cloud controlplane to open a port of the flow."""
+    app_id = os.getenv("LIGHTNING_CLOUD_APP_ID", None)
+    project_id = os.getenv("LIGHTNING_CLOUD_PROJECT_ID", None)
+
+    if not app_id or not project_id:
+        raise Exception("The app_id and project_id should be defined.")
+
+    assert project_id
```
Comment (ethanwharris) [inline src/lightning_app/utilities/port.py:64]:
```suggestion
```

File: src/lightning_app/utilities/port.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+    client = LightningClient()
+    list_apps_resp = client.lightningapp_instance_service_list_lightningapp_instances(project_id=project_id)
+    lit_app: Optional[Externalv1LightningappInstance] = None
+
+    for lightningapp in list_apps_resp.lightningapps:
+        if lightningapp.id == app_id:
+            lit_app = lightningapp
+
+    assert lit_app
```
Comment (ethanwharris) [inline src/lightning_app/utilities/port.py:74]:
```suggestion
    if not lit_app:
        raise RuntimeError(
            "App was not found. Please open an issue at https://github.com/lightning-AI/lightning/issues."
        )
```

File: src/lightning_app/utilities/port.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+def disable_port(port: int, ignore_disabled: bool = True) -> None:
+    """Make a request to the cloud controlplane to close a port of the flow."""
+
+    app_id = os.getenv("LIGHTNING_CLOUD_APP_ID", None)
+    project_id = os.getenv("LIGHTNING_CLOUD_PROJECT_ID", None)
+
+    if not app_id or not project_id:
+        raise Exception("The app_id and project_id should be defined.")
+
+    assert project_id
```
Comment (ethanwharris) [inline src/lightning_app/utilities/port.py:107]:
```suggestion
```

File: src/lightning_app/utilities/port.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+    client = LightningClient()
+    list_apps_resp = client.lightningapp_instance_service_list_lightningapp_instances(project_id=project_id)
+    lit_app: Optional[Externalv1LightningappInstance] = None
+
+    for lightningapp in list_apps_resp.lightningapps:
+        if lightningapp.id == app_id:
+            lit_app = lightningapp
+
+    assert lit_app
```
Comment (ethanwharris) [inline src/lightning_app/utilities/port.py:117]:
```suggestion
    if not lit_app:
        raise RuntimeError(
            "App was not found. Please open an issue at https://github.com/lightning-AI/lightning/issues."
        )
```

**Review by ethanwharris — APPROVED** [review]
Looks good, just a lot of assertions haha

**Pull request merged**